### PR TITLE
feat(EMS-870): Account sign in - API validation, OTP/security code generation

### PR DESCRIPTION
**Pull request by ttbarnes**
This PR adds the ability to verify exporter account sign in credentials against the database and then generate a OTP/security code.

## Changes

- New custom resolver/mutation: `accountSignIn`. This mutation:
  - Gets the account from DB.
  - Checks if the password is valid.
  - Generates a OTP/security code with `otplib` package.
  - Sends an email with OTP/security code.
- Add new function to send "here is your security code" email.
- UI: Add new GQL api call for sign in.
- UI: Update the sign in controller to call sign in GQL call. Add a temporary redirect to `invalid-credentials` (this will be replaced with a re-render in a future PR).
- E2E: 
  - Create cypress commands `getExporterByEmail`, and `verifyAccountEmail`.
  - Update some tests so that sign in functionality is using a verified account (happy path).

## Other improvements
- Extract `createAccount` mutation into it's own file instead of being inside custom-schemaa
- Extract "get an account by X field" function into it's own helper file.
- Update `verifyAccountEmailAddress` resolver/mutation to consume new "get an account by X field" function, simplifying the function, DRY.
- Rename `PASSWORD` constants to `ENCRYPTION` since it's now used in multiple places and it's not exclusive to "account password".

### CODE REVIEW

File: src/api/.keystone/config.js [hunk 1]
```diff
@@ -69,19 +69,29 @@ var ACCOUNT = {
...
+  ENCRYPTION: {
     RANDOM_BYTES_SIZE: 32,
     STRING_TYPE: "hex",
     PBKDF2: {
       ITERATIONS: 1e4,
       KEY_LENGTH: 64,
       DIGEST_ALGORITHM: "sha512"
     }
+  },
+  OTP: {
```
Comment (ttbarnes) [inline src/api/.keystone/config.js:81]: ⚠️ NOTE: this file is auto generated/compiled by keystone.

File: e2e-tests/cypress/support/insurance/account/verify-account-email.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+/**
+ * verifyAccountEmail
+ * Get the exporter (with verification hash) directly from the API,
+ * Mimic "cliking email verification link" in an email inbox by manually navigating to the URL
+ */
+const verifyAccountEmail = () => {
+  // get exporter
+  cy.getExporterByEmail(exporterEmail).then((response) => {
+    const { data } = response.body;
```
Comment (abhi-markan) [inline e2e-tests/cypress/support/insurance/account/verify-account-email.js:17]: Should we consider `response` and `data` void check, otherwise please wrap the in try-catch.

File: e2e-tests/cypress/support/insurance/account/get-exporter-by-email.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
+import api from '../../api';
+
+/**
+ * getExporterByEmail
+ * @param {String} Exporter account email
+ * Get an exporter directly from the API,
+ * @returns {Object} Exporter account
+ */
+const getExporterByEmail = (email) => {
+  api.getExporterByEmail(email).then((response) => {
```
Comment (abhi-markan) [inline e2e-tests/cypress/support/insurance/account/get-exporter-by-email.js:10]: Wrap them in try-catch please.

File: src/api/helpers/generate-otp.ts [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    const expiry = OTP.VERIFICATION_EXPIRY();
+
+    return {
+      securityCode,
+      salt,
+      hash,
+      expiry,
+    };
+  } catch (err) {
+    throw new Error(`Error generating OTP ${err}`);
```
Comment (abhi-markan) [inline src/api/helpers/generate-otp.ts:42]: Please add console.error.

File: src/api/helpers/is-valid-account-password.test.ts [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+      const result = isValidAccountPassword(mockPassword, salt, hash);
+
+      expect(result).toEqual(true);
+    });
+  });
+
+  describe('when password is invalid', () => {
+    it('should return false', () => {
+      const result = isValidAccountPassword('IncorrectPassword123!', salt, hash);
+
```
Comment (abhi-markan) [inline src/api/helpers/is-valid-account-password.test.ts:37]: Please add a test case where password provided in empty or will this be prevented by TS and UI validation?

File: src/ui/server/controllers/insurance/account/sign-in/index.ts [hunk 6]
```diff
@@ -83,6 +84,23 @@ export const post = (req: Request, res: Response) => {
     });
   }
 
-  // TODO: check sign in details are correct and generate a code.
-  return res.redirect(ENTER_CODE);
+  try {
+    // validate credentials
```
Comment (abhi-markan) [inline src/ui/server/controllers/insurance/account/sign-in/index.ts:88]: Can we please add `req.body` void check.

**Review by abhi-markan — CHANGES_REQUESTED** [review]
Great work, few comments to address please.

File: src/api/helpers/is-valid-account-password.test.ts (hunk 5 above)
Comment (ttbarnes) [reply src/api/helpers/is-valid-account-password.test.ts:37]: This will be prevented by TS on the API and UI level - the functions cannot be called without all parameters being provided.

File: src/ui/server/controllers/insurance/account/sign-in/index.ts (hunk 6 above)
Comment (ttbarnes) [reply src/ui/server/controllers/insurance/account/sign-in/index.ts:88]: This is done by TS, a `req` will always have a body

File: e2e-tests/cypress/support/insurance/account/verify-account-email.js (hunk 2 above)
Comment (ttbarnes) [reply e2e-tests/cypress/support/insurance/account/verify-account-email.js]: response/data check is handled by `cy.getExporterByEmail` so no need for it here as well - i've added try/catch to both.

**Review by abhi-markan — APPROVED** [review]
